### ISSUE TITLE
Add support for a lastModified option for some methods.

### ISSUE THREAD

**Issue by n1k0** [Kinto/kinto-client#32]
Following up #27 where I realize the `safe` option always needs a `last_modified` value passed to work at all :D

Methods to update:

- [ ] `Bucket#createCollection`œ
- [ ] `Bucket#setPermissions`
- [ ] `Collection#setPermissions`
- [ ] `Collection#setMetadata`
- [ ] `Collection#setSchema`
- [ ] `Collection#deleteRecord`

Notes:

- `Collection#updateRecord` already check for the `last_modified` value of the record, if any;
- `batch` implementations should deal with inner calls to these methods as well... Might be easier to require a resource object passed with a `last_modified` value...

So that'd be either:

```js
bucket("blog").updateCollection({id: "posts", last_modified: 1234567890}, {safe: true});
``` 

Or

```js
bucket("blog").updateCollection("posts", {lastModified: 1234567890, safe: true});
``` 

We could also drop the `safe` Boolean option and suppose we want concurrency checks as soon as a `lastModified` option is passed (and then the second version would make a lot of sense, as it would be super explicit). Though not obvious how that would work with `batch`...

Last, maybe `lastModified` wouldn't be the best possible name for this option... Maybe something more explicit like `ifNotModifiedSince`, thought that's rather long.

Feedback welcome.

**Comment by almet** [comment]
The option doesn't **always** need the `lastModified` option.

The `safe` option, in my head ([and in the kinto.py client](https://github.com/Kinto/kinto.py/blob/master/kinto_client/__init__.py#L26)) can also do a "do not override": it will check if the id doesn't exist before writing it (when doing a put for instance), using the http header `If-None-Match: *`.

When passing the `lastModified` option, [it should actually have the headers specified](https://github.com/Kinto/kinto.py/blob/master/kinto_client/__init__.py#L125-L126).

Hope this helps.

**Comment by n1k0** [comment]
Yeah, true. So keeping the `safe` option makes sense, and what you describe with the do not override is the current behavior, forgot about it.

So keeping the `safe` option makes sense. I'm just trying to wrap my head around what would be the most convenient API for handling every intents:

- Fire and forget: do not check anything
- Ensure nothing is overriden: safe option passed as true
- Ensure nothing have been updated meanwhile: safe option passed as true AND lastModified sent for comparison

Translating this with an easy to use API is far from trivial, so any dream API is warmly welcome here.

**Comment by n1k0** [comment]
Maybe `safe` could be either a Boolean or a timestamp? Thats probably sounds weird though that would allow us writing things like this:

```js
// Fire and forget
bucket("blog").updateCollection("posts");
// Do not override
bucket("blog").updateCollection("posts", {safe: true});
// Concurrency check
bucket("blog").updateCollection("posts", {safe: 1234567890});
```

I'm not easy with multiple types like this, so if you have a better idea, I'm all ears :)

**Comment by n1k0** [comment]
Note that requiring an object to be always passed, as for `Collection#updateRecord` would cleanly solve the problem:

```js
// Fire and forget
bucket("blog").updateCollection({id: "posts"});
// Do not override
bucket("blog").updateCollection({id: "posts"}, {safe: true});
// Concurrency check
bucket("blog").updateCollection({id: "posts", last_modified: 1234567890}, {safe: true});
```

Though that's not very convenient to use methinks. Thoughts?

**Comment by n1k0** [comment]
Now I think about it a little more, we could also move the collection update operation at the Collection level itself:

 ```js
// Fire and forget
bucket("blog").collection("posts").update({schema});
// Do not override
bucket("blog").collection("posts").update({schema, safe: true});
// Concurrency check
bucket("blog").collection("posts").update({schema, safe: true}); // checks for last_modified
```

But that would mean the Collection object should always know about its last_modified, and as this requires an HTTP request, that means an async operation and so a Promise, so that API wouldn't work; that would be rather something like this:

 ```js
// Fire and forget
bucket("blog").collection("posts").then(coll => coll.update({schema}));
```

Which is a little appalling if you ask me:

 ```js
// Fire and forget
bucket("blog")
  .then(bucket => bucket.collection("posts"))
  .then(coll => coll.update({schema}));
```

Here Bucket and Collection resolved instances could know about their respective `last_modified` value, but come on, that's not user friendly at all... No?

**Comment by Natim** [comment]
What we did for Kinto.py is to have the ``safe`` parameter as well as the ``last_modified`` one and default the ``last_modified`` value to the record ``last_modified`` value if present.

``safe: true`` is the default which means that the lines below are equivalent:

    bucket("blog").updateCollection({id: "posts", last_modified: 1234567890});
    bucket("blog").updateCollection({id: "posts", last_modified: 1234567890}, {safe: true});
    bucket("blog").updateCollection({id: "posts"}, {safe: true, last_modified: 1234567890});
    bucket("blog").updateCollection({id: "posts", last_modified: 1234567890},
                                    {safe: true, last_modified: 1234567890});

So the only case you would want to explicitely write are the following:

    bucket("blog").deleteCollection({id: "posts"}, {last_modified: 1234567890});
    bucket("blog").deleteCollection({id: "posts"}, {safe: false});

Also note that: ``bucket("blog").createCollection({id: "posts"});`` will have the ``If-None-Match: *`` header.

**Comment by n1k0** [comment]
Note: For Collection and Bucket lastModified property, we could fetch it at first call, and cache it locally, updating that value each time we modify it. That means a supplementary request for each bucket, and each collection though.

Another way to limit the impact on performances would be to preload all buckets timestamps using `KintoClient#getBucket()` the first time we perform an operation on it, like what we do for fecthing server settings. The same could be done for collections.

Sounds a little bloated though.

**Comment by n1k0** [comment]
@Natim: I realize that always passing an object instead of a string identifier would be the most obvious way of solving all the situations we describe, as you suggested in your example. I think I'll have to update these signatures accordingly.

**Comment by n1k0** [comment]
Fixed in #37.
